### ISSUE TITLE
Sampled MTP with penalties: a request co-admitted with another diverges from the exact eager sampler, and the lockstep gates no longer compare live transcripts

### ISSUE THREAD

**Issue by HawkBearPig** [HawkBearPig/dgpp#15]
Found while reviewing #9 (the contributor's record mentions "a batched transcript comparison discrepancy" they left alone — this is that, reproduced and narrowed). Two separate things, both pre-existing in `master` (present at `444441f`, unchanged by #9, which has since been merged).

### 1. The sampled-MTP lockstep gates stopped comparing live transcripts

`glm_tp_serving_mtp_graph_sampling_matches_eager_speculator` (and, until #9 was merged, `run_mtp_depth_sampling_gate`) follow a request through `sched.find(id)`: they compare `res->generated` with the eager speculator's stream after every tick and stop following the request once `res->status != kActive`. Since the request-history retention change, a stored `Result` is only filled in at retirement: while a request is live, `generated` is empty and `status` is still the default `kQueued`. So after the first tick the harness marks every request finished, the eager oracle takes exactly one step per request, and the only comparison that ever runs is against an empty vector.

Evidence: printing both sides at the end of the unmodified gate (it still reports `[ OK ]`):

```
req solo  graph [41 90 81 44 77 11 64]  oracle-stream [41]  next 90
req a     graph [41 57 17 4 21 62 18]   oracle-stream [41]  next 57
req b     graph [12 4 24 51 91 46]      oracle-stream [12]  next 56     <- already disagrees
```

#9 (now on `master`) fixed this for the depth gate by following live output through a `SchedulerObserver`; the depth-1 gate above (and anything else using the `find()` pattern) still needs the same treatment.

### 2. With the comparison made real: a sampled request with penalties is not reproducible next to a co-admitted request

Setup (the gate's own): world 2, MTP depth 1, 9-token prompt, `temperature 1.0`, `top_p 0.95`, `frequency_penalty 0.2`, `presence_penalty 0.1`, candidate cap 24 (every step takes the exact host fallback), request `a` seed 11 / 7 steps, request `b` seed 13 / 6 steps, identical prompts.

| how `b` runs | graph engine | eager sampled speculator |
|---|---|---|
| alone (phases `{a}`, then `{b}`) | `12 56 38 63 70 5` | `12 56 38 63 70 5` |
| admitted in the same tick as `a`, two slots | `12 4 24 51 91 46` | `12 56 …` |
| same, `batch_min_live` above the slot count (scalar graphs only) | `12 4 …` | `12 56 …` |
| same, penalties set to 0 | matches the oracle | — |

`a` (slot 0) matches the oracle in every row; it is always the second request. The prefill pick (`12`) agrees; the first decode token does not. Identical on both ranks, so the ranks agree with each other — it is the graph engine against the eager oracle.

What that rules in and out:

- not the batched graph (the scalar-graph run diverges the same way);
- not PR #9 (its engine change only moves the acceptance counters; the divergence is there with or without it);
- it needs the penalties: with `frequency_penalty`/`presence_penalty` at 0 the concurrent run follows the oracle for both requests. That points at the penalty context for the second request of a grouped admission — the device count table / context mirror for slot 1 (seeded from the prompt and the prefill pick at admission), rather than the RNG or the draft.

Impact as far as I can tell: requests at the served defaults (no penalties) are unaffected. A request that sets a presence/frequency penalty and is admitted together with another request does not get the exact seeded sample the eager path produces — a reproducibility/exactness problem, not a crash. Not yet checked: `repetition_penalty`, depth 2, whether a second slot that was admitted a tick *later* shows it too, and which of the two sides is the one that is wrong (the eager oracle is self-consistent between the alone and concurrent cases, which is why I suspect the graph side).

### Reproduction

On `master` (since #9 it has the observer-based harness), in `run_mtp_depth_sampling_gate` in `tests/cuda/glm_tp_test.cpp`: set `kSlots = 2` for depth 1 and run `a` and `b` in one phase (as depth 2 already does), then

```
DGPP_TEST_FILTER=depth1_graph_sampling ./build-ci/glm_tp_test
[FAIL] ...: request 'b' graph transcript differs from the eager sampled speculator's after replay 6
```

Running the two requests in separate phases, or zeroing the two penalties, makes the transcript checks pass.

**Comment by HawkBearPig** [comment]
Fixed in [b13d06314f24](https://github.com/HawkBearPig/dgpp/commit/b13d06314f241cdedb96125633265718aff66762), pushed to `master`.

The live-observer repair reproduced the reported mismatch on both ranks. Penalty counts and the saved verifier logits were correct. The first request's fallback drained the whole batch, publishing the second request's next draft before its verdict was collected. Its host fallback then tested draft 44 even though the verifier had consumed draft 19.

The adapter now preserves every request's verified drafts before collecting any current verdict or draining that replay. These buffers are preallocated and reused. The disabled-pipeline path also waits for the verdict before draining, avoiding a read from an emptied replay queue. Model arithmetic and sampling rules are unchanged.

The old sampling gate now follows token and retirement events and checks each complete transcript. Coverage includes seed 13 alone/co-admitted, captured-slot reuse, staggered admission, fallback acceptance counters, and `DGPP_PIPELINE=0`. Both depth gates assert the actual scalar/batch replay counters. The above-slot-count threshold mentioned in the report is clamped, so it did not disable batching; depth two provides a verified scalar co-admission control.

Validation:

- Full CI/release builds, all 30 host tests, and GLM, GLM-4.7, Qwen and DeepSeek graph-engine suites passed. The expanded depth-one oracle passed 43 forced fallbacks with pipelining both on and off.
- Four-node real-model checks: all six solo-versus-concurrent comparisons matched tokens and logprobs exactly, covering frequency/presence, repetition, and no penalties. Actual two-slot replay counts were checked.
- All 18 measured greedy control responses and usage records matched baseline. Median decode throughput changed by +0.03% to +0.35%, within measurement noise.
- All nine API checks passed and operation streams matched across all four ranks.

[Full reproduction and validation record](https://github.com/HawkBearPig/dgpp/blob/b13d06314f241cdedb96125633265718aff66762/benchmarks/results/2026-09-21-mtp-fallback-isolation.md).

Production is now running the clean `b13d06314f24` release on all four nodes. Final API checks passed; `/proc/PID/exe` hashes and operation-stream hashes match across ranks, and the deployment configuration is unchanged.
